include/seqan3/io/alignment_file/format_bam.hpp

```diff
@@ -0,0 +1,1061 @@
+        alignment_record_core core;
+        std::ranges::copy_n(stream_view.begin(), sizeof(core), reinterpret_cast<char *>(&core));
+
+        if (core.refID >= static_cast<int32_t>(header.ref_ids().size()) || core.refID < -1) // [[unlikely]]
+            throw format_error{detail::to_string("Reference id index '", core.refID, "' is not in range of ",
+                                                 "header.ref_ids(), which has size ", header.ref_ids().size(), ".")};
+        else if (core.refID > -1) // not unmapped
+            ref_id = core.refID;                                                   // field::REF_ID
+
+        flag = core.flag;                                                          // field::FLAG
+        mapq = core.mapq;                                                          // field::MAPQ
+
+        if (core.pos > -1) // [[likely]]
+            ref_offset = core.pos;                                                 // field::REF_OFFSET
+
+        if constexpr (!detail::decays_to_ignore_v<mate_type>)                      // field::MATE
+        {
+            if (core.next_refID > -1)
+                get<0>(mate) = core.next_refID;
+
+            if (core.next_pos > -1) // [[likely]]
+                get<1>(mate) = core.next_pos;
+
+            get<2>(mate) = core.tlen;
+        }
+
+        // read id
+        // -------------------------------------------------------------------------------------------------------------
+        read_field(stream_view | view::take_exactly_or_throw(core.l_read_name - 1), id); // field::ID
+        std::ranges::next(std::ranges::begin(stream_view)); // skip '\0'
+
+        // read cigar string
+        // -------------------------------------------------------------------------------------------------------------
+        if constexpr (!detail::decays_to_ignore_v<align_type>)
+        {
+            std::tie(cigar, ref_length, seq_length, offset_tmp, soft_clipping_end) =
+                detail::parse_binary_cigar(stream_view, core.n_cigar_op);
+        }
+        else
+        {
+            detail::consume(stream_view | view::take_exactly_or_throw(core.n_cigar_op * 4));
+        }
+
+        offset = offset_tmp;
+
+        // read sequence
+        // -------------------------------------------------------------------------------------------------------------
+        if (core.l_seq > 0) // sequence information is given
+        {
+            auto seq_stream = stream_view
+                            | view::take_exactly_or_throw(core.l_seq / 2) // one too short if uneven
+                            | std::view::transform([] (char c) -> std::pair<sam_dna16, sam_dna16>
+                              {
+                                  return {sam_dna16{}.assign_rank(std::min(15, static_cast<uint8_t>(c) >> 4)),
+                                          sam_dna16{}.assign_rank(std::min(15, static_cast<uint8_t>(c) & 0x0f))};
+                              });
```

Comment:
The sequence of BAM is packed: 2 symbols into 1 byte. 
`>> 4` retrieves the upper 4 bits and `&0x0f` the lower 4 bits.
Then I ensure that the rank never exceeds 15 to avoid UB and assign this rank to sam_dna16 and return a pair of the two alphabet symbols.